### PR TITLE
Update BuildingTeams PySol content/3_Silver/Graph_Traversal.mdx

### PR DESCRIPTION
**Pull request by CryoJS**
### Description
Adds new Python Solutions to https://cses.fi/problemset/task/1668/ for content/3_Silver/Graph_Traversal.mdx.
Relates to #3444

### Runtimes
DFS (missing Python): https://cses.fi/problemset/result/11118878/ (0.74s)
Previous BFS Sol: https://cses.fi/problemset/result/11118911/ (0.90s)
New BFS Sol: https://cses.fi/problemset/result/11118910/ (0.69s)

_Note: Updated/changed the pre-existing Python BFS solution as I had a simpler, faster solution._

---

_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [X] I have tested my code.
- [X] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [X] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [X] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: content/3_Silver/Graph_Traversal.mdx [hunk 1]
```diff
@@ -1082,6 +1082,47 @@ public class BuildingTeams {
 ```
 
 </JavaSection>
+<PySection>
+
+Note that the below solution passes in CPython3 but exceeds the time limit on test case #7 in PyPy3.
```
Comment (SansPapyrus683) [inline content/3_Silver/Graph_Traversal.mdx:1087]: wrap in warning and just say you have to submit w/ cpython

File: content/3_Silver/Graph_Traversal.mdx [hunk 2]
```diff
@@ -1082,6 +1082,47 @@ public class BuildingTeams {
...
+	adj[a - 1].append(b - 1)
+	adj[b - 1].append(a - 1)
+
+
+def dfs(node):
+	for next_node in adj[node]:
+		if team[next_node]:
+			if team[next_node] == team[node]:
+				print("IMPOSSIBLE")
+				raise SystemExit
```
Comment (freakin23) [inline content/3_Silver/Graph_Traversal.mdx:1111]: u can also use exit(0)

File: content/3_Silver/Graph_Traversal.mdx (hunk 1 above)
Comment (CryoJS) [reply content/3_Silver/Graph_Traversal.mdx:1087]: Got it!

File: content/3_Silver/Graph_Traversal.mdx (hunk 2 above)
Comment (CryoJS) [reply content/3_Silver/Graph_Traversal.mdx:1111]: Yep, thanks. I like to use SystemExit just due to preference lol.

Although I do think it is slightly faster and more descriptive. https://stackoverflow.com/a/35430500

**Review by ryanchou-dev — APPROVED** [review]

File: content/3_Silver/Graph_Traversal.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply content/3_Silver/Graph_Traversal.mdx:1113]: wait this is the first time i've seen this used what

File: content/3_Silver/Graph_Traversal.mdx [hunk 3]
```diff
@@ -1082,6 +1082,49 @@ public class BuildingTeams {
...
+<PySection>
+
+<Warning>
+You have to submit with CPython (not PyPy3) to avoid exceeding the time limit.
+</Warning>
+
+```py
+import sys
+
+input = sys.stdin.readline
```
Comment (SansPapyrus683) [inline content/3_Silver/Graph_Traversal.mdx:1094]: is just using `input()` too slow

File: content/3_Silver/Graph_Traversal.mdx [hunk 4]
```diff
@@ -1082,6 +1082,49 @@ public class BuildingTeams {
...
+
+sys.setrecursionlimit(int(1e5 + 1))
```
Comment (SansPapyrus683) [inline content/3_Silver/Graph_Traversal.mdx:1096]: magic #

File: content/3_Silver/Graph_Traversal.mdx [hunk 5]
```diff
@@ -1082,6 +1082,49 @@ public class BuildingTeams {
...
+adj = [[] for _ in range(n)]
+team = [0] * n  # 0 -> has not been assigned a team yet
+
+for _ in range(m):
+	a, b = map(int, map(int, input().strip().split()))
+	adj[a - 1].append(b - 1)
+	adj[b - 1].append(a - 1)
+
+
+def dfs(node):
```
Comment (SansPapyrus683) [inline content/3_Silver/Graph_Traversal.mdx:1108]: type annotations

File: content/3_Silver/Graph_Traversal.mdx [hunk 6]
```diff
@@ -1210,46 +1253,38 @@ public class BuildingTeams {
 <PySection>
 
 ```py
+import sys
 from collections import deque
 
-n, m = map(int, input().split())
+input = sys.stdin.readline
```
Comment (SansPapyrus683) [inline content/3_Silver/Graph_Traversal.mdx:1259]: same thing here

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/3_Silver/Graph_Traversal.mdx (hunk 3 above)
Comment (CryoJS) [reply content/3_Silver/Graph_Traversal.mdx:1094]: Yep, just tested it

https://cses.fi/problemset/result/11126765/ fails
https://cses.fi/problemset/result/11126767/ passes

File: content/3_Silver/Graph_Traversal.mdx (hunk 6 above)
Comment (CryoJS) [reply content/3_Silver/Graph_Traversal.mdx:1259]: It barely passes in the BFS version. Should I include or exclude fast input?
Fast Input: https://cses.fi/problemset/result/11118910/
No Fast Input: https://cses.fi/problemset/result/11126773/

File: content/3_Silver/Graph_Traversal.mdx (hunk 4 above)
Comment (CryoJS) [reply content/3_Silver/Graph_Traversal.mdx:1096]: For some reason, n + 3 is the magic number

Comment (SansPapyrus683) [reply content/3_Silver/Graph_Traversal.mdx]: would you mind setting it to 10 ** 9
like add a comment "disable recursion limit"

File: content/3_Silver/Graph_Traversal.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply content/3_Silver/Graph_Traversal.mdx:1113]: ok so for consistency with rest of code could you use exit()

File: content/3_Silver/Graph_Traversal.mdx [hunk 7]
```diff
@@ -1210,46 +1255,38 @@ public class BuildingTeams {
...
+team = [0] * n  # 0 -> has not been assigned a team yet
+
 for _ in range(m):
-	a, b = map(int, input().split())
+	a, b = map(int, map(int, input().strip().split()))
 	adj[a - 1].append(b - 1)
 	adj[b - 1].append(a - 1)
 
-assigned = [0 for _ in range(n)]
-valid = True
```
Comment (SansPapyrus683) [inline content/3_Silver/Graph_Traversal.mdx:1224]: just now realizing that this is replacing an old sol
uneasy about this bc the old one was consistent w/ java and c++, while this one isn't

Comment (CryoJS) [reply content/3_Silver/Graph_Traversal.mdx:1224]: > _Note: Updated/changed the pre-existing Python BFS solution as I had a simpler, faster solution._

Yeah, I mentioned it in the description. If that's a concern, we just just use the old one. If so, I can try to see if there are any optimizations that can be made while still keeping it consistent with the other language's solutions (unless it's exactly the same).

Comment (SansPapyrus683) [reply content/3_Silver/Graph_Traversal.mdx:1224]: > I can try to see if there are any optimizations that can be made while still keeping it consistent

i'd love this

**Comment by CryoJS** [comment]
Yep, of course!

**Comment by CryoJS** [comment]
Here are the newest submissions:
DFS -> https://cses.fi/problemset/result/11193528/ (AC, slowest is 0.73s)
BFS -> https://cses.fi/problemset/result/11193531/ (AC, slowest is 0.90s)

I couldn't find any optimizations, the code is pretty much exactly the same as the C++ code.
Nevertheless, how they managed to make BFS run slower than DFS, I don't know lol.

**Review by SansPapyrus683 — APPROVED** [review]
image

**Pull request merged**